### PR TITLE
[Merged by Bors] - feat: port Topology.Order.LowerTopology

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+variable [CompleteLattice α] [CompleteLattice β] [TopologicalSpace α] [LowerTopology α]
+  [TopologicalSpace β] [LowerTopology β]
+
+theorem InfₛHom.continuous (f : InfₛHom α β) : Continuous f := by
+  refine LowerTopology.continuous_of_Ici fun b => ?_
+  convert LowerTopology.isClosed_Ici (infₛ <| f ⁻¹' Ici b)
+  refine' Subset.antisymm (fun a => infₛ_le) fun a ha => le_trans _ <| OrderHomClass.mono f ha
+  simp [map_infₛ]
+#align Inf_hom.continuous InfₛHom.continuous
```
Comment (Vierkantor) [inline Mathlib/Topology/Order/LowerTopology.lean:281]: [The explanation of the typeclass issues on zulip.](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Instances.20in.20!4.232138/near/326644735)

Here's one way to hint to Lean which instance we want. Note that we're actually casting `f` to an `OrderHom` now, and `simp` is having the same trouble synthesising the `OrderHomClass` instance so we can't simp `((f : α →ₒ β) : α → β)` down to just `f : α → β`, so I use `refine` instead.
```suggestion
theorem InfₛHom.continuous (f : InfₛHom α β) : Continuous f := by
  refine LowerTopology.continuous_of_Ici fun b => ?_
  convert LowerTopology.isClosed_Ici (infₛ <| f ⁻¹' Ici b)
  refine' Subset.antisymm (fun a => infₛ_le) fun a ha => le_trans _ <|
    OrderHomClass.mono (f : α →o β) ha
  refine' LE.le.trans _ (map_infₛ f _).ge
  simp
#align Inf_hom.continuous InfₛHom.continuous
```

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+theorem toLower_ofLower (a : WithLowerTopology α) : toLower (ofLower a) = a :=
+  rfl
+#align with_lower_topology.to_lower_of_lower WithLowerTopology.toLower_ofLower
+
+@[simp]
+theorem ofLower_toLower (a : α) : ofLower (toLower a) = a :=
+  rfl
+#align with_lower_topology.of_lower_to_lower WithLowerTopology.ofLower_toLower
+
+@[simp]
```
Comment (urkud) [inline Mathlib/Topology/Order/LowerTopology.lean:96]:
```suggestion
-- porting note: removed @[simp] to make linter happy
```

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+theorem toLower_inj {a b : α} : toLower a = toLower b ↔ a = b :=
+  Iff.rfl
+#align with_lower_topology.to_lower_inj WithLowerTopology.toLower_inj
+
+@[simp]
```
Comment (urkud) [inline Mathlib/Topology/Order/LowerTopology.lean:101]:
```suggestion
-- porting note: removed @[simp] to make linter happy
```

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+  refine' Subset.antisymm (fun a => infₛ_le) fun a ha => le_trans _ <|
+    OrderHomClass.mono (f : α →o β) ha
+  refine' LE.le.trans _ (map_infₛ f _).ge
+  simp
+#align Inf_hom.continuous InfₛHom.continuous
+
+-- see Note [lower instance priority]
+instance (priority := 90) LowerTopology.to_continuousInf : ContinuousInf α :=
+  ⟨(infInfₛHom : InfₛHom (α × α) α).continuous⟩
+#align lower_topology.to_has_continuous_inf LowerTopology.to_continuousInf
```
Comment (mcdoll) [inline Mathlib/Topology/Order/LowerTopology.lean:288]:
```suggestion
instance (priority := 90) LowerTopology.continuousInf : ContinuousInf α :=
  ⟨(infInfₛHom : InfₛHom (α × α) α).continuous⟩
#align lower_topology.to_has_continuous_inf LowerTopology.continuousInf
```
and then it should be also changed in the module docstring.

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+end Preorder
+
+section PartialOrder
+
+variable [PartialOrder α] [TopologicalSpace α] [LowerTopology α]
+
+-- see Note [lower instance priority]
+/-- The lower topology on a partial order is T₀. -/
+instance (priority := 90) toSpace : T0Space α :=
```
Comment (mcdoll) [inline Mathlib/Topology/Order/LowerTopology.lean:246]: Shouldn't the name be `t0Space`?

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+/-- Type synonym for a preorder equipped with the lower topology. -/
+def WithLowerTopology := α
+#align with_lower_topology WithLowerTopology
+
+variable {α β}
+
+namespace WithLowerTopology
+
+/-- `to_lower` is the identity function to the `WithLowerTopology` of a type.  -/
```
Comment (jcommelin) [inline Mathlib/Topology/Order/LowerTopology.lean:65]:
```suggestion
/-- `toLower` is the identity function to the `WithLowerTopology` of a type.  -/
```

File: Mathlib/Topology/Order/LowerTopology.lean [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+@[match_pattern]
+def toLower : α ≃ WithLowerTopology α := Equiv.refl _
+#align with_lower_topology.to_lower WithLowerTopology.toLower
+
+/-- `of_lower` is the identity function from the `WithLowerTopology` of a type.  -/
```
Comment (jcommelin) [inline Mathlib/Topology/Order/LowerTopology.lean:70]:
```suggestion
/-- `ofLower` is the identity function from the `WithLowerTopology` of a type.  -/
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**